### ISSUE TITLE
incorrect detection of functions/classes as variables in cppcoreguidelines-avoid-non-const-global-variables

### ISSUE THREAD

**Issue by aminya** [llvm/llvm-project#47527]
|  |  |
| --- | --- |
| Bugzilla Link | [48183](https://llvm.org/bz48183) |
| Version | unspecified |
| OS | All |

## Extended Description 
cppcoreguidelines-avoid-non-const-global-variables incorrectly detects top level functions as variables.

This is the type of error:
```
 warning: variable 'Test0' is non-const and globally accessible, consider making it const [cppcoreguidelines-avoid-non-const-global-variables]
```

This is a code to reproduce the issue:

test.h

```cpp
#include <optional>

const auto cond = false;
std::optional<int> Test0() {
    if (cond) {
        return 1;
    }
    return {};
}

```
or

test.h
```cpp
class Test1 {
  int a;
}

class Test2: public Test1 {
  int b;
}
```

**Comment by carlosgalvezp** [comment]
No longer reproducible on trunk.
